### PR TITLE
Updates to Tpsa discretization

### PR DESCRIPTION
**Pull request by keileg**
## Proposed changes
This PR provides the following changes:
* An update of the Tpsa discretization, after the discretization scheme was redefined.
* Improved test coverage for Tpsa, including consistency testing for 3d problems.
* Fixes to the boundary condition discretization identified due to the improved tests.

Integration of Tpsa into a multiphysics framework will be provided in an upcoming PR.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [x] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [x] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tests/numerics/fv/test_tpsa.py [hunk 1]
```diff
@@ -1,9 +1,24 @@
 """Tests for the Tpsa discretization. Contains two sets of tests:
-1. TestTpsaTailoredGrid is a class that defines a cell of two grids and verifies the
-   discretization matrices for the Tpsa discretization. Both interior cells and all
-   types of boundary conditions are tested.
-2. Various other tests that probe different aspects of the Tpsa discretization.
-   These are not grouped in a class, but are individual tests.
+1. Various other tests that probe different aspects of the Tpsa discretization. These
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:2]:
```suggestion
1. Various tests that probe different aspects of the Tpsa discretization. These
```

File: tests/numerics/fv/test_tpsa.py [hunk 2]
```diff
@@ -1,9 +1,24 @@
...
+   check that the resulting primary variables have the expected signs.
+   - test_translation: Assign uniform Dirichlet condition, check that the material is
+   translated as expected and that it is stress free.
+   - test_robin_neumann_dirichlet_consistency: Send the Robin parameter to 0 and towards
+   infinity, check consistency with Dirichlet and Neumann conditions.
+   - test_3d_linear_displacement: Check that the discretization is exact for a 3d
+   material that undergoes linear displacement.
+
+2. Detailed tests of the implemented discretization scheme, with hard-coded values for
+   the various discretization matrices defined on an 2x2 grid, contained in the class
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:14]:
```suggestion
   the various discretization matrices defined on a 2x2 grid, contained in the class
```

File: tests/numerics/fv/test_tpsa.py [hunk 3]
```diff
@@ -21,1418 +36,1426 @@
...
+@pytest.mark.parametrize("driving_bc_type", ["dir", "neu"])
+@pytest.mark.parametrize("tensile", [False, True])
+def test_compression_tension(g: pp.Grid, driving_bc_type: str, tensile: bool):
+    """Assign a compressive or tensile force, check that the resulting displacement is
+    in the correct direction, and that the total pressure is negative.
 
-    Parameters:
-        grid: Grid to discretize.
-        d: Dictionary with parameters.
+    The logic of the test is as follows (assuming a compressive regime, tensile=False,
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:46]: Please revisit this paragraph, or risk me sending you this https://xkcd.com/859/ once a day until the parentheses are closed.

Comment (IvarStefansson) [reply tests/numerics/fv/test_tpsa.py:46]: Please also consider spelling out the type of condition. Both "compressive" and "neutral" could be either dir or neu.

File: tests/numerics/fv/test_tpsa.py [hunk 4]
```diff
@@ -21,1418 +36,1426 @@
...
 
-    d[pp.PARAMETERS][KEYWORD]["bc"] = bc_disp
-    if include_rot:
-        d[pp.PARAMETERS][KEYWORD]["bc_rot"] = bc_rot
+    if g.dim == 2:
+        rot_dim = 1
+    else:
+        rot_dim = g.dim
 
+    # Boundary values, map to right-hand side.
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:97]: I don't get this comment.

File: tests/numerics/fv/test_tpsa.py [hunk 5]
```diff
@@ -21,1418 +36,1426 @@
...
+    in the correct direction, and that the total pressure is negative.
 
-    Parameters:
-        grid: Grid to discretize.
-        d: Dictionary with parameters.
+    The logic of the test is as follows (assuming a compressive regime, tensile=False,
+    is compressive on the south, east and (if 3d) bottom faces, the remaining faces
+    having neutral conditions. Check that this results in displacement in the negative
+    x-direction and positive y- (and z) direction. The total pressure should be
+    negative, while EK cannot surmise the correct sign of the rotation by physical
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:50]: Doesn't the sign of total pressure depend on tensile?

File: tests/numerics/fv/test_tpsa.py [hunk 6]
```diff
@@ -21,1418 +36,1426 @@
...
+    d = _set_uniform_parameters(g)
 
-        # Define shear modulus and Cosserat parameter.
-        self.mu_0 = 1
-        self.mu_1 = 2
-        self.cos_0 = 1
-        self.cos_1 = 3
+    # Set type and values of boundary conditions. No rotation.
+    _set_uniform_bc(g, d, "dir")
+    if g.dim == 2:
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:130]: You could just use the 3d in both cases.

File: tests/numerics/fv/test_tpsa.py [hunk 7]
```diff
@@ -21,1418 +36,1426 @@
...
-        self.d_1_6 = 3 / (2 * self.n_6_nrm)
+    # Check: Uniform translation should result in a stress-free state on all faces. This
+    # is not the case for rotation and solid mass, which are only zero when integrated
+    # over individual cells.
+    disp_cells = np.zeros(g.num_cells * g.dim)
+    disp_cells[:: g.dim] = disp[0]
+    disp_cells[1 :: g.dim] = disp[1]
+    if g.dim == 3:
+        disp_cells[2 :: g.dim] = disp[2]
+    v = matrices["stress"] @ disp_cells + matrices["bound_stress"] @ bc_values.ravel(
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:160]: Rename variable?

File: tests/numerics/fv/test_tpsa.py [hunk 8]
```diff
@@ -21,1418 +36,1426 @@
...
-        # identified by prefix c_. On Dirichlet faces, the face value is imposed, thus
-        # the cell is asigned weight 0 (hence the complement has weight 1).
-        c2f_avg_0 = 0
-        c2f_avg_6 = 0
-        c_c2f_avg_0 = 1 - c2f_avg_0
-        c_c2f_avg_6 = 1 - c2f_avg_6
+    # For future reference: EK has verified that, as the Robin weight is increased, the
+    # discretization matrices of the Dirichlet and Robin cases do converge. The rates
+    # seem to be faster for the matrices than for the solution, but this is not
+    # unreasonable.
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:224]: Why not?

File: tests/numerics/fv/test_tpsa.py [hunk 9]
```diff
@@ -21,1418 +36,1426 @@
...
+    Neumann (controlled by the parameter neu_bcs). In both cases, the bottom boundary is
+    assigned homogenous Dirichlet conditions to ensure solvability.
 
-        bound_rotation_diffusion = np.zeros((2, 7))
-        bound_rotation_diffusion[0, 0] = -1
-        bound_rotation_diffusion[1, 6] = 1
+    """
+    # Notes for debugging: For Neumann BCs, at least two cells are needed in the x- or
+    # y-direction to fix the domain and ensure solvability.
+    g = pp.CartGrid([2, 1, 10], [1, 1, 1])
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:277]: Is run time negligible? Otherwise, consider 
```suggestion
    g = pp.CartGrid([2, 1, 4], [1, 1, 1])
```

File: tests/numerics/fv/test_tpsa.py [hunk 10]
```diff
@@ -21,1418 +36,1426 @@
...
-        bound_displacement_solid_pressure_cell[1, 0] = (
-            -self.n_0[1] * c_c2f_avg_0_y / ((2 * mu_0_d + rw_0_y) * self.n_0_nrm)
-        )
-        bound_displacement_solid_pressure_cell[2, 1] = (
-            self.n_6[0] * c_c2f_avg_6 / ((2 * mu_1_d + rw_6) * self.n_6_nrm)
-        )
-        bound_displacement_solid_pressure_cell[3, 1] = (
-            self.n_6[1] * c_c2f_avg_6 / ((2 * mu_1_d + rw_6) * self.n_6_nrm)
-        )
+    flux = sps.block_array(
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:392]: Why do you call this "flux"?

File: tests/numerics/fv/test_tpsa.py [hunk 11]
```diff
@@ -21,1418 +36,1426 @@
...
-        bound_displacement_solid_pressure_cell = np.zeros((4, 2))
-        bound_displacement_solid_pressure_cell[1, 0] = (
-            -self.d_0_0 * self.n_0[1] / (2 * self.mu_0 * self.n_0_nrm)
-        )
-        bound_displacement_solid_pressure_cell[3, 1] = (
-            self.d_1_6 * self.n_6[1] / (2 * self.mu_1 * self.n_6_nrm)
-        )
+    The boundary condition has value 0.1 on the south face, -0.1 on the east face, and
+    0.1 on the bottom face (if 3d). The remaining faces have homogeneous conditions.
+    The conditions correspond to a compressive force on the south, east and (if
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:543]: Whether the condition is compressive or tensile depends on the bc values, no?

File: tests/numerics/fv/test_tpsa.py [hunk 12]
```diff
@@ -21,1418 +36,1426 @@
...
-    The logic of the test is as follows (assuming a compressive regime, tensile=False,
-    is compressive on the south, east and (if 3d) bottom faces, the remaining faces
-    having neutral conditions. Check that this results in displacement in the negative
-    x-direction and positive y- (and z) direction. The total pressure should be
-    negative, while EK cannot surmise the correct sign of the rotation by physical
-    intuition. The grid is assumed to be Cartesian (or else the discretization is
-    inconsistent, and anything can happen), in which case the face normal vectors will
-    point into the domain on the south (and bottom) boundary, out of the domain on the
-    east boundary.
+# See module-level comment on the class TestTpsaTailoredGrid and on the motivation for
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:603]: Not reviewed past this point.

**Review by IvarStefansson — APPROVED** [review]
Lot's of stuff I had to take on faith. However, I think I managed to write one or two useful comments. And certainly some useless ones. Feel free to merge once addressed.

File: src/porepy/numerics/fv/tpsa.py [hunk 13]
```diff
@@ -723,46 +664,43 @@ def discretize(self, sd: Grid, data: dict) -> None:
...
-            weights=dist.mu_by_dist_fc_cc,
-            minlength=nf,
-        )
+        # Arithmetic average of shear modulus.
+        arithmetic_average_shear_modulus = dist.arithmetic_average_shear_modulus
         # Following the paper, we filter away Dirichlet boundary conditions.
         mass_total_pressure = -filters.dir_notpass @ (
             sps.dia_matrix(
-                (sd.face_areas / (2 * arithmetic_average_shear_modulus), 0),
+                (sd.face_areas / arithmetic_average_shear_modulus, 0),
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:672]: Yes, it is included in the arithmetic_average_shear_modulus (with appropriate adjustments for Robin boundaries). Sigh.

File: src/porepy/numerics/fv/tpsa.py [hunk 14]
```diff
@@ -723,46 +664,43 @@ def discretize(self, sd: Grid, data: dict) -> None:
...
         if nd == 3:
-            # In this case, \hat{R}_k^n = \bar{R}_k^n is the 3x3 matrix given in the
+            # In this case, \hat{R}_k^n = \bar{R}_k^n = R^n is the 3x3 matrix given in the
             # Tpsa paper,
             #
-            #    R^n = [[0, -n2, n0], [n2, 0, -n0], [-n1, n0, 0]]
+            #    R^n = [[0, -n2, n1], [n2, 0, -n0], [-n1, n0, 0]]
             #
             # However, for efficient implementation we will use a utility function for
             # matrix construction available in PorePy which, it turns out, requires a
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:706]: Please clarify the comment, there is no mention of Neumann here.

File: tests/numerics/fv/test_tpsa.py (hunk 5 above)
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:50]: Correct. Specified 'negative in compression'.

File: tests/numerics/fv/test_tpsa.py (hunk 4 above)
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:97]: It made no sense. Improved.

File: tests/numerics/fv/test_tpsa.py (hunk 6 above)
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:130]: Thanks.

File: tests/numerics/fv/test_tpsa.py (hunk 7 above)
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:160]: Done.

File: tests/numerics/fv/test_tpsa.py (hunk 8 above)
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:224]: Outdated comment, deleted. Sorry.

File: tests/numerics/fv/test_tpsa.py (hunk 10 above)
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:392]: Habit: First block row is stress, second is face rotations, third is actually flux of solid mass. Scalings that alter the physical interpretation may apply.

Renamed to face_quantities for lack of a better expression.

File: tests/numerics/fv/test_tpsa.py (hunk 11 above)
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:543]: This function sets a compressive regime, while tensile can be achieved by negation in the calling function.This is now explicitly noted (last line of this paragraph).
